### PR TITLE
streamingccl: switch PartitionedStreamClient to use pgx library

### LINKED ISSUE #83694: streamingccl: data race from Subscription.Subscribe
**Describe the problem**

When attempting to remove `skip.UnderRace` from `TestPartitionedStreamReplicationClient`, I hit:

```
==================
WARNING: DATA RACE
Write at 0x00c00146e328 by goroutine 714:
  runtime.slicecopy()
      GOROOT/src/runtime/slice.go:284 +0x0
  bufio.(*Reader).Read()
      GOROOT/src/bufio/bufio.go:238 +0x6c4
  io.ReadAtLeast()
      GOROOT/src/io/io.go:328 +0xdd
  io.ReadFull()
      GOROOT/src/io/io.go:347 +0x324
  github.com/lib/pq.(*conn).recvMessage()
      github.com/lib/pq/external/com_github_lib_pq/conn.go:1018 +0x2bd
  github.com/lib/pq.(*conn).recv1Buf()
      github.com/lib/pq/external/com_github_lib_pq/conn.go:1059 +0x3a
  github.com/lib/pq.(*rows).Next()
      github.com/lib/pq/external/com_github_lib_pq/conn.go:1573 +0x20e
  github.com/lib/pq.(*rows).Close()
      github.com/lib/pq/external/com_github_lib_pq/conn.go:1530 +0xc4
  database/sql.(*Rows).close.func1()
      GOROOT/src/database/sql/sql.go:3280 +0x57
  database/sql.withLock()
      GOROOT/src/database/sql/sql.go:3396 +0xb5
  database/sql.(*Rows).close()
      GOROOT/src/database/sql/sql.go:3279 +0x24d
  database/sql.(*Rows).awaitDone()
      GOROOT/src/database/sql/sql.go:2933 +0x10b
  database/sql.(*Rows).initContextClose·dwrap·27()
      GOROOT/src/database/sql/sql.go:2917 +0x71

Previous read at 0x00c00146e328 by goroutine 478:
  runtime.slicecopy()
      GOROOT/src/runtime/slice.go:284 +0x0
  database/sql.cloneBytes()
      GOROOT/src/database/sql/convert.go:494 +0x38c
  database/sql.convertAssignRows()
      GOROOT/src/database/sql/convert.go:261 +0x357
  database/sql.(*Rows).Scan()
      GOROOT/src/database/sql/sql.go:3246 +0x526
  github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient.(*partitionedStreamSubscription).Subscribe.func1()
      github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:304 +0x104
  github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient.(*partitionedStreamSubscription).Subscribe()
      github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:316 +0x4a1
  github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient.Subscription.Subscribe-fm()
      github.com/cockroachdb/cockroach/pkg/ccl/streamingccl/streamclient/client.go:113 +0x6b
  github.com/cockroachdb/cockroach/pkg/util/ctxgroup.Group.GoCtx.func1()
      github.com/cockroachdb/cockroach/pkg/util/ctxgroup/ctxgroup.go:169 +0x4c
  golang.org/x/sync/errgroup.(*Group).Go.func1()
      golang.org/x/sync/errgroup/external/org_golang_x_sync/errgroup/errgroup.go:74 +0x91
```

This looks to me like the race is actually in the `lib/pq` driver in which it is accessing the rows buffer in its asynchronous context cancellation handling at the same time as accessing it from `Scan`.

See also:

- https://github.com/lib/pq/issues/1021
- https://go-review.googlesource.com/c/go/+/89936/
- https://github.com/lib/pq/pull/706

Jira issue: CRDB-17212

### PR DESCRIPTION
**Pull request by gh-casper**
We have been switching to pgx from lib/pq across teams in CRL.
This PR continues this journey and makes PartitionedStreamClient 
 to use pgx and also in a thread-safe way.

To avoid overhead of creating connection, APIs except Subscribe
share the same connection guarded by a mutex, and Subcribe owns
a private connection which makes sense as two Subscribe calls
don't want to share a connection.

Fixes data race we have been seen in the past: 
Closes: https://github.com/cockroachdb/cockroach/issues/83694

Also unskip stream_replication_e2e_test under race.

Release note: None

Addresses https://github.com/cockroachdb/cockroach/issues/83695

### CODE REVIEW

File: pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go [hunk 1]
```diff
@@ -306,12 +304,10 @@ func TestTenantStreamingProducerJobTimedOut(t *testing.T) {
...
 	defer log.Scope(t).Close(t)
 
-	// TODO(casper): now this has the same race issue with
-	// TestPartitionedStreamReplicationClient, please fix them together in the future.
-	skip.UnderRace(t, "disabled under race")
-
 	ctx := context.Background()
-	c, cleanup := createTenantStreamingClusters(ctx, t, defaultTenantStreamingClustersArgs)
+	args := defaultTenantStreamingClustersArgs
+	args.srcClusterSettings[`stream_replication.job_liveness_timeout`] = `'20s'`
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go:309]: I suspect that might be too low for race tests;
Please verify that your test passes w/ `./dev test --stress --race`

File: pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go [hunk 2]
```diff
@@ -19,35 +18,43 @@ import (
...
 	}
 }
 
 func newPartitionedStreamClient(remote *url.URL) (*partitionedStreamClient, error) {
-	db, err := gosql.Open("postgres", remote.String())
+	config, err := pgx.ParseConfig(remote.String())
+	if err != nil {
+		return nil, err
+	}
+	conn, err := pgx.ConnectConfig(context.Background(), config)
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:47]: Do we want to use background ctx here? Seems like piping ctx is not much of a lift; 
I don't know what pgx does in the face of context cancellation; but it definitely does nothing w/ background context.

File: pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go [hunk 3]
```diff
@@ -60,19 +67,16 @@ func (p *partitionedStreamClient) Create(
 	ctx, sp := tracing.ChildSpan(ctx, "streamclient.Client.Create")
 	defer sp.Finish()
 
-	streamID := streaming.InvalidStreamID
-
-	conn, err := p.srcDB.Conn(ctx)
+	p.mu.Lock()
+	defer p.mu.Unlock()
+	var streamID streaming.StreamID
+	row := p.mu.srcConn.QueryRow(ctx, `SELECT crdb_internal.start_replication_stream($1)`, tenantID.ToUint64())
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:73]: why do we need to keep srcConn under mu? is it not thread safe?

File: pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go [hunk 4]
```diff
@@ -274,16 +265,23 @@ func (p *partitionedStreamSubscription) Subscribe(ctx context.Context) error {
 	defer sp.Finish()
 
 	defer close(p.eventsChan)
-	conn, err := p.db.Conn(ctx)
+	// Each subscription has its own pgx connection instead of
+	// shares one with other Subscription.
```
Comment (miretskiy) [inline pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go:269]: drop instead of ...? or rephrase.

**Review by gh-casper — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/84394)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @samiskin, and @stevendanna)

___
*[`pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go` line 47 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84394#-N6tVr3wGcF9bv2om6TG-r2-47:-N7qNSGc135AV5zITosC:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/448899fbd41f5c441cddf2438236201cbf8148ee/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go#L47)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

Do we want to use background ctx here? Seems like piping ctx is not much of a lift; 
I don't know what pgx does in the face of context cancellation; but it definitely does nothing w/ background context.
</blockquote></details>

Done.
___
*[`pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go` line 73 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84394#-N6tVr3wGcF9bv2om6TG-r2-73:-N7qNVBfEpZNX8wLe_vx:biql732) ([raw file](https://github.com/cockroachdb/cockroach/blob/448899fbd41f5c441cddf2438236201cbf8148ee/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go#L73)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

why do we need to keep srcConn under mu? is it not thread safe?
</blockquote></details>

No, it's not thread-safe. Creating a private connection on each call is expensive, so using a shared conn.
___
*[`pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go` line 269 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84394#-N6tVr3wGcF9bv2om6TG-r2-269:-N7qO4rqExk-gVFdW2F2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/448899fbd41f5c441cddf2438236201cbf8148ee/pkg/ccl/streamingccl/streamclient/partitioned_stream_client.go#L269)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

drop instead of ...? or rephrase.
</blockquote></details>

Done.



_Code quote:_
```Go
Each subscription has its own pgx connection instead of
```
___
*[`pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go` line 309 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/84394#-N6tVr3x4XfGX4jwj0oI-r2-309:-N7qZRyS8fe7t7NAl7-1:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/448899fbd41f5c441cddf2438236201cbf8148ee/pkg/ccl/streamingccl/streamingest/stream_replication_e2e_test.go#L309)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

I suspect that might be too low for race tests;
Please verify that your test passes w/ `./dev test --stress --race`
</blockquote></details>

Done.



_Code quote:_
```Go
job_liveness_timeout
```

**Review by miretskiy — APPROVED** [review]

**Pull request merged**